### PR TITLE
Adding module to process text using AWS Polly text-to-speech engine

### PR DESCRIPTION
**Pull request by slapula**
##### SUMMARY
This module utilizes the AWS Polly text-to-speech engine to process text files and convert them into spoken audio.  You can process either local text files or text files hosted on S3.  You can also output the resulting audio file locally or to S3 as well.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
`aws_polly_speech`

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = None
  configured module search path = [u'/Users/asmith/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/asmith/.pyenv/versions/2.7.13/lib/python2.7/site-packages/ansible
  executable location = /Users/asmith/.pyenv/versions/2.7.13/bin/ansible
  python version = 2.7.13 (default, Sep  8 2017, 15:16:38) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    path: 'migration_instructions.mp3'
+'''
+
+
+RETURN = r'''#'''
+
+from contextlib import closing
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, AWSRetry
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: These are unused.

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict, boto3_tag_list_to_ansible_dict
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: As are these.

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        try:
+            response = s3_client.get_object(
+                Bucket=module.params.get('src_bucket'),
+                Key=module.params.get('src_path')
+            )
+            params['Text'] = response['Body'].read().decode("utf-8") 
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json_aws(e, msg="Failed to get text file from S3")
+
+    create_synthesis(client, s3_client, module, params, result)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: It looks like changed=True is always the case. You don't need to use/pass around a result dict. `module.exit_json(changed=True)`

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = r'''
+---
+module: polly_speech
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: Maybe `aws_polly_speech`

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    - If you specify ssml as the `text_type`, follow the SSML format for the input text.
+    choices: ['local', 's3']
+    required: true
+  src_bucket:
+    description:
+    - Specifies the S3 bucket that stores the text file.
+  src_path:
+    description:
+    - Specifies the local path or S3 key path where the text file resides.
+  destination:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: Maybe instead an additional parameter to specify the destination, make dst_bucket and dst_path mutually exclusive and require one of them.
```
AnsibleAWSModule(...,
                 required_one_of=[('dst_bucket', 'dst_path')],
                 mutually_exclusive=[('dst_bucket', 'dst_path')])
```

File: lib/ansible/modules/cloud/amazon/polly_speech.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+        ('source', 'local', ['src_path']),
+        ('source', 's3', ['src_bucket', 'src_path']),
+        ('destination', 'local', ['dst_path']),
+        ('destination', 's3', ['dst_bucket', 'dst_path'])
+    ]
+
+    module = AnsibleAWSModule(
+        argument_spec={
+            'lexicons': dict(type='list'),
+            'output_format': dict(type='str', choices=['json', 'mp3', 'ogg_vorbis', 'pcm'], required=True),
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/polly_speech.py]: type='str' is the default so you can omit all of those.

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by s-hertel** [comment]
This is neat :-)

File: lib/ansible/modules/cloud/amazon/polly_speech.py (hunk 5 above)
Comment (slapula) [reply lib/ansible/modules/cloud/amazon/polly_speech.py]: Maybe I should make this a little more clear in the documentation but both `src_path` and `dst_path` have shared responsibility.  When the source and/or destination are local, the `_path` parameter functions as the absolute or relative path to the text file on the filesystem.  When the source and/or destination are S3, `_path` parameter functions as the S3 key in relation to the bucket.  To make a long story short, `_path` will always be used but `_bucket` will only be needed when S3 is designated as the source or destination.

**Comment by mattclay** [comment]
CI permissions update needed:

```
Traceback (most recent call last):
  File "/tmp/ansible_5ufT8m/ansible_module_aws_polly_speech.py", line 111, in create_synthesis
    response = client.synthesize_speech(**params)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
ClientError: An error occurred (AccessDeniedException) when calling the SynthesizeSpeech operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=68339.66 is not authorized to perform: polly:SynthesizeSpeech
```

File: lib/ansible/modules/cloud/amazon/aws_polly_speech.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    - Specifies whether the input text is plain text or SSML.
+    choices: ['ssml', 'text']
+    default: 'text'
+  voice_id:
+    description:
+    - Voice ID to use for the synthesis.
+    choices: ['Geraint', 'Gwyneth', 'Mads', 'Naja', 'Hans', 'Marlene', 'Nicole', 'Russell', 'Amy', 'Brian', 'Emma', 'Raveena', 'Ivy', 'Joanna',
+        'Joey', 'Justin', 'Kendra', 'Kimberly', 'Matthew', 'Salli', 'Conchita', 'Enrique', 'Miguel', 'Penelope', 'Chantal', 'Celine', 'Mathieu',
+        'Dora', 'Karl', 'Carla', 'Giorgio', 'Mizuki', 'Liv', 'Lotte', 'Ruben', 'Ewa', 'Jacek', 'Jan', 'Maja', 'Ricardo', 'Vitoria', 'Cristiano',
+        'Ines', 'Carmen', 'Maxim', 'Tatyana', 'Astrid', 'Filiz', 'Vicki', 'Takumi', 'Seoyeon', 'Aditi']
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/aws_polly_speech.py:51]: This seems like something that will change frequently, is there an API method to list these so we can do validation at run-time and not stop people from using new voices as they become available.

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
Hi @slapula, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.aws<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>

**Pull request closed without merging**